File: javascript/011.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(heights) {
  let i = 0;
  let j = heights.length - 1;

  const getVolume = (i, j) => {
    const height = heights[i] < heights[j] ? heights[i] : heights[j];
    return height * (j - i);
  }

  let maxVolume = getVolume(i, j);
  while(true) {
    const iHeight = heights[i];
    const jHeight = heights[j];
    if (iHeight < jHeight) {
      while(iHeight > heights[++i]) {}
    } else {
      while(jHeight > heights[--j]) {}
    }
    if (i >= j) break;
    const volume = getVolume(i, j);
    if (volume > maxVolume) maxVolume = volume;
  }

  return maxVolume;
};

File: ruby/011.rb
# @param {Integer[]} height
# @return {Integer}

def get_volumn(height, i, j)
  length = j - i
  if height[i] < height[j]
    length * height[i]
  else
    length * height[j]
  end
end

def max_area(height)
  i = 0
  j = height.length - 1

  max_volumn = get_volumn(height, i, j)
  while i < j do
    if height[i] < height[j]
      i_height = height[i]
      i += 1 while height[i] <= i_height && i < j
    else
      j_height = height[j]
      j -= 1 while height[j] <= j_height && i < j
    end

    break if i >= j

    new_volumn = get_volumn(height, i, j)
    max_volumn = new_volumn if new_volumn > max_volumn
  end

  max_volumn
end